port/osx/omrosdump.c

```diff
@@ -0,0 +1,473 @@
+
+		segments[i].fileoff = seg_file_off;
+		written = pwrite(corefile_fd, &segments[i], sizeof(struct segment_command_64), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing segment commands:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+
+		kr = mach_vm_read(task_port, segments[i].vmaddr, segments[i].vmsize, &data_read, &data_size);
+		if (KERN_SUCCESS != kr) {
+			mach_error("mach_vm_read on memory segment failed with: ", kr);
+			goto done;
+		}
+		written = pwrite(corefile_fd, (void *)data_read, data_size, seg_file_off);
+		if (written < 0) {
+			perror("pwrite() error writing segment data:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		mach_vm_deallocate(task_port, data_read, data_size);
+
+		file_off += sizeof(struct segment_command_64);
+		seg_file_off += segments[i].vmsize;
+	}
+	mh64.ncmds += segment_count;
+	mh64.sizeofcmds += segment_count * sizeof(struct segment_command_64);
+
+	/* write mach header after all command number and size are known */
+	written = pwrite(corefile_fd, &mh64, sizeof(struct mach_header_64), 0);
+	if (written < 0) {
+		perror("pwrite() error writing mach header:");
+		kr = KERN_FAILURE;
+		goto done;
+	}
+
+done:
+	if (corefile_fd > 0) {
+		close(corefile_fd);
+		if (KERN_SUCCESS != kr) {
+			remove(corefile_name);
+		}
+	}
+	return kr;
+}
+
+static kern_return_t
+list_thread_commands(mach_port_t task_port, struct thread_command_full_64 **thread_commands, natural_t *thread_count)
+{
+	kern_return_t kr = KERN_SUCCESS;
+	thread_act_array_t thread_info;
+	struct thread_command_full_64 *threads = NULL;
+	natural_t i = 0;
+
+	kr = task_threads(task_port, &thread_info, thread_count);
+
+	if (KERN_SUCCESS != kr) {
+		mach_error("task_thread failed with: ", kr);
+		return kr;
+	}
+
+	threads = calloc(*thread_count, sizeof(struct thread_command_full_64));
+	if (NULL == threads) {
+		kr = KERN_NO_SPACE;
+		goto done;
+	}
+
+	for (i = 0; i < *thread_count; i++) {
+		uint32_t state_int_count = 0;
+		threads[i].cmd = LC_THREAD;
+		threads[i].cmdsize = sizeof(uint32_t) + sizeof(uint32_t); /* sizes of the cmd and cmdsize fields */
+		state_int_count = x86_THREAD_STATE_COUNT;
+		kr = thread_get_state(thread_info[i], x86_THREAD_STATE,
+			(thread_state_t)&threads[i].thread_state, &state_int_count);
+		if (KERN_SUCCESS != kr) {
+			goto done;
+		}
+		threads[i].cmdsize += state_int_count * sizeof(thread_state_t);
+		state_int_count = x86_FLOAT_STATE_COUNT;
+		kr = thread_get_state(thread_info[i], x86_FLOAT_STATE,
+			(thread_state_t)&threads[i].float_state, &state_int_count);
+		if (KERN_SUCCESS != kr) {
+			goto done;
+		}
+		threads[i].cmdsize += state_int_count * sizeof(thread_state_t);
+		state_int_count = x86_EXCEPTION_STATE_COUNT;
+		kr = thread_get_state(thread_info[i], x86_EXCEPTION_STATE,
+			(thread_state_t)&threads[i].exceptions, &state_int_count);
+		if (KERN_SUCCESS != kr) {
+			goto done;
+		}
+		threads[i].cmdsize += state_int_count * sizeof(thread_state_t);
```

Comment:
I think these occurrences of `sizeof(thread_state_t)` should be `sizeof(uint32_t)`.